### ISSUE TITLE
Can't find C Solver

### ISSUE THREAD

**Issue by scopatz** [ergs/transmutagen-paper#5]
New issue during make is that it can't seem to find the C solver.  I have pointed the TRANSMUTAGEN variable to both the source directory and the installed directory and it produces basically the same error message in both cases.

```console
Traceback (most recent call last):
  File "/home/scopatz/miniconda/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/scopatz/miniconda/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/scopatz/.local/lib/python3.6/site-packages/transmutagen-1.1.0+83.g16f1549-py3.6.egg/transmutagen/analysis.py", line 1129, in <module>
    analyze()
  File "/home/scopatz/.local/lib/python3.6/site-packages/transmutagen-1.1.0+83.g16f1549-py3.6.egg/transmutagen/analysis.py", line 1112, in analyze
    degree=args.degree)
  File "/home/scopatz/.local/lib/python3.6/site-packages/transmutagen-1.1.0+83.g16f1549-py3.6.egg/transmutagen/analysis.py", line 265, in analyze_nofission
    file=file, title=title)
  File "/home/scopatz/.local/lib/python3.6/site-packages/transmutagen-1.1.0+83.g16f1549-py3.6.egg/transmutagen/analysis.py", line 292, in plot_nofission_transmutes
    m = nofission_transmutes[time_name][lib][r]
KeyError: 'transmutagen generated C solver'
make: *** [Makefile:48: nofission-pwru50-1-day.pgf] Error 1
```

**Comment by scopatz** [comment]
CC @asmeurer

**Comment by asmeurer** [comment]
Ah yeah. You have to build the py_solve module for this to work. I guess I should add some code to the Makefile to make sure it is built.

**Comment by asmeurer** [comment]
In the meantime, run 

```
python -m transmutagen.gensolve --py-solve
cd py_solve
python setup.py build_ext --inplace
```

and point `TRANSMUTAGEN` to the root of the git directory (it gets added to PYTHONPATH [here](https://github.com/ergs/transmutagen-paper/blob/4f93f479fc3dd411f06890d681c9c6b2f3c71e28/Makefile#L7)). Or alternately, you can install py_solve globally.

**Comment by asmeurer** [comment]
Pull the transmutagen and paper repo and try again.

**Comment by scopatz** [comment]
I am still seeing this error, but I will try to install. Is there a way that we can build out-of-place?

**Comment by scopatz** [comment]
```console
scopatz@artemis ~/transmutagen-paper master $ env TRANSMUTAGEN_DATA=~/transmutagen-data TRANSMUTAGEN=~/.local/lib/python3.6/site-packages/transmutagen-1.1.0+87.g9f1ad6a-py3.6.egg/transmutagen make
python cram_plot.py
python error_plot.py
if [ ! -d "/home/scopatz/.local/lib/python3.6/site-packages/transmutagen-1.1.0+87.g9f1ad6a-py3.6.egg/transmutagen/py_solve" ]; then echo "py_solve not found. If the following errors, set the TRANSMUTAGEN environment variable to the root of the transmutagen git repo."; fi
py_solve not found. If the following errors, set the TRANSMUTAGEN environment variable to the root of the transmutagen git repo.
python -m transmutagen.gensolve --py-solve -o /home/scopatz/.local/lib/python3.6/site-packages/transmutagen-1.1.0+87.g9f1ad6a-py3.6.egg/transmutagen/py_solve/py_solve/solve.c
Computing coefficients for degree 6
Computing coefficients for degree 8
Computing coefficients for degree 10
Computing coefficients for degree 12
Computing coefficients for degree 14
Computing coefficients for degree 16
Computing coefficients for degree 18
Traceback (most recent call last):
  File "/home/scopatz/miniconda/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/scopatz/miniconda/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/scopatz/.local/lib/python3.6/site-packages/transmutagen-1.1.0+87.g9f1ad6a-py3.6.egg/transmutagen/gensolve.py", line 738, in <module>
    main()
  File "/home/scopatz/.local/lib/python3.6/site-packages/transmutagen-1.1.0+87.g9f1ad6a-py3.6.egg/transmutagen/gensolve.py", line 734, in main
    generate(**arguments)
  File "/home/scopatz/.local/lib/python3.6/site-packages/transmutagen-1.1.0+87.g9f1ad6a-py3.6.egg/transmutagen/gensolve.py", line 670, in generate
    write_if_diff(outfile, src)
  File "/home/scopatz/.local/lib/python3.6/site-packages/transmutagen-1.1.0+87.g9f1ad6a-py3.6.egg/transmutagen/gensolve.py", line 526, in write_if_diff
    with open(filename, 'w') as f:
FileNotFoundError: [Errno 2] No such file or directory: '/home/scopatz/.local/lib/python3.6/site-packages/transmutagen-1.1.0+87.g9f1ad6a-py3.6.egg/transmutagen/py_solve/py_solve/solve.c'
make: *** [Makefile:49: nofission-pwru50-1-day.pgf] Error 1
scopatz@artemis ~/transmutagen-paper master $ 
```

**Comment by scopatz** [comment]
Note that even after doing, 

```
python -m transmutagen.gensolve --py-solve
cd py_solve
python setup.py install --user
```

I still receive the same build error.

**Comment by asmeurer** [comment]
You need to point TRANSMUTAGEN to the git directory, as explained in the error message (I guess I should have made it red). py_solve is not installed with transmutagen. 

I can remove the auto building and make it so that you have to make sure it is built and updated manually, but then it will work if you install it globally.

**Comment by scopatz** [comment]
Right, so I installed it globally and it failed in the same manner as when I pointed it to the global install of transmutagen.  I don't want to point it to the repo and I don't want to hava an in place build because that implies that transmutagen is not really viable as a package. I'd like to be able to do everything out-of-source, assuming I have installed transmutagen and py_solve globally.

**Comment by asmeurer** [comment]
I'm thinking that py_solve should just be part of transmutagen, as a submodule. The downside of this is that you have to build it when you install transmutagen (including regenerating the solver each time, to make sure it hasn't changed). I'll look into doing this. Hopefully the it isn't too complicated to get setuptools to do what I want.

**Comment by scopatz** [comment]
Yeah, that would be better.  You could perhaps have it so that py_solve is only optionally installed.

**Comment by asmeurer** [comment]
OK, I have moved it inside transmutagen. You should be able to just inplace build or install transmutagen and it should work (as always, be sure to build both `transmutagen` and `transmutagen-paper`). I'm still waiting on Travis and a full paper rebuild to verify that I didn't miss anything, so let me know if you find any issues.

**Comment by asmeurer** [comment]
I didn't make it optional. I think enough code in transmutagen uses it that it's worth having (anyway, making it optional was more work, so I didn't bother with it, but if it's important we can add it). 

The downside of this is that setup.py now has to import transmutagen to generate the solver. It also cannot know if the generator needs updating, so it has to regenerate the solver every time you run setup.py, which can take a minute. It is smart enough to not recompile the solver if it hasn't changed, though.

**Comment by asmeurer** [comment]
Also I didn't bother adding Clang support, because I don't know how to check what compiler is being used in setup.py.

**Comment by scopatz** [comment]
> Also I didn't bother adding Clang support, because I don't know how to check what compiler is being used in setup.py.

Yeah I think you have to check the env var CC

**Comment by scopatz** [comment]
I am now seeing the following issue after successfully installing transmutagen *(which does install py_solve, as you say)

```console
scopatz@artemis ~/transmutagen-paper master $ env TRANSMUTAGEN_DATA=~/transmutagen-data TRANSMUTAGEN=~/.local/lib/python3.6/site-packages/transmutagen-1.1.0+89.g13bb03c-py3.6-linux-x86_64.egg/transmutagen/ make 
python -m transmutagen.analysis --nofission  --file nofission.pgf
analyzing /home/scopatz/.local/lib/python3.6/site-packages/transmutagen-1.1.0+89.g13bb03c-py3.6-linux-x86_64.egg/transmutagen/tests/data/pwru50_400000000000000.0_nofission.npz on 1 day
Traceback (most recent call last):
  File "/home/scopatz/miniconda/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/scopatz/miniconda/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/scopatz/.local/lib/python3.6/site-packages/transmutagen-1.1.0+89.g13bb03c-py3.6-linux-x86_64.egg/transmutagen/analysis.py", line 1130, in <module>
    analyze()
  File "/home/scopatz/.local/lib/python3.6/site-packages/transmutagen-1.1.0+89.g13bb03c-py3.6-linux-x86_64.egg/transmutagen/analysis.py", line 1113, in analyze
    degree=args.degree)
  File "/home/scopatz/.local/lib/python3.6/site-packages/transmutagen-1.1.0+89.g13bb03c-py3.6-linux-x86_64.egg/transmutagen/analysis.py", line 263, in analyze_nofission
    alpha0=alpha0)
TypeError: run_transmute_test() got an unexpected keyword argument 'require_py_solve'
make: *** [Makefile:48: nofission-pwru50-1-day.pgf] Error 1
```

**Comment by asmeurer** [comment]
Ah crap. Pull and try again.

**Comment by scopatz** [comment]
Woohoo! It finally builds!

**Comment by asmeurer** [comment]
Great. It looks like the dist building is broken, so I won't be able to release until I fix that.

**Comment by scopatz** [comment]
uggh sorry

**Comment by asmeurer** [comment]
The sdist isn't including the pyx file (only the generated .c file). But the setup.py is set up to Cythonize when it installs (https://travis-ci.org/ergs/transmutagen/jobs/356063272#L1858). 

Any thoughts how this should be fixed? I've never really shipped Cythonized code before. Should I add something to make it include the pyx file, or should I modify setup.py somehow to not cythonize if the .c files are already there? Also is it really correct to have `cythonize` be called at the top-level of setup.py? Is there no way to make it only be called when running `build`?

**Comment by scopatz** [comment]
I'd add  "*.pyx" and "*.pxd" file to the `"package_data"` for the transmutagen package and the `transmutagen.py_solve` subpackage.  Here is the pyne example that just adds the pxds: https://github.com/pyne/pyne/blob/develop/setup_sub.py#L50 

I would NOT add the generated C files because these can and will changes based on the cython version and potenailly the platforms.  

Hopefully this helps!

**Comment by asmeurer** [comment]
The .c files seem to be adding themselves. I'm not sure how to remove them. I guess if I include the pyx file it won't matter because it will just get regenerated.

**Comment by scopatz** [comment]
right
